src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedRecordsWriterFactoryTest.java

```diff
@@ -71,147 +79,89 @@ public void constructor_createsDirectoryIfNotExists() throws Exception {
 
     // Assert
-    assertThat(writer.getPath())
-        .isEqualTo(
-            new Path(
-                targetDirectoryPath.getFileSystem(conf).resolvePath(targetDirectoryPath),
-                "dwhassessment_2022-12-25T120022.3344_a665f132.avro"));
-    assertThat(fs.exists(targetDirectoryPath)).isTrue();
-    assertThat(existedBefore).isFalse();
+    assertThat(getDirectoryFiles(fs, targetDirectoryPath)).isEmpty();
   }
 
   @Test
-  public void getWriter_updatesDirectoryWithRollover() throws Exception {
+  public void getWriter_writeEventsToTheSameWriterIfWithinRolloverWindow() throws Exception {
     TickableFixedClock clock = new TickableFixedClock(DEFAULT_TIMESTAMP);
+    LocalDate targetDate = DEFAULT_TIMESTAMP.atOffset(ZoneOffset.UTC).toLocalDate();
+    Path targetDirectoryPath = new Path(tmpFolder, targetDate.toString());
+    FileSystem fs = targetDirectoryPath.getFileSystem(conf);
     DatePartitionedRecordsWriterFactory datePartitionedRecordsWriterFactory =
         createRecordsWriterFactory(clock);
-    // Act
-    RecordsWriter writer1 = datePartitionedRecordsWriterFactory.createWriter();
-    clock.tick(Duration.ofMinutes(20));
-    RecordsWriter writer2 = datePartitionedRecordsWriterFactory.createWriter();
-    clock.tick(Duration.ofDays(1));
-    datePartitionedRecordsWriterFactory.maybeUpdateRolloverTime();
-    RecordsWriter writer3 = datePartitionedRecordsWriterFactory.createWriter();
-
-    // Assert
-    assertThat(writer1.getPath().getName()).isEqualTo(createExpectedFileName(DEFAULT_TIMESTAMP));
-    Instant secondRecordInstant = parseDateTime("2022-12-25T12:20:22.3344Z");
-    assertThat(writer2.getPath().getName()).isEqualTo(createExpectedFileName(secondRecordInstant));
-    assertThat(writer2.getPath().getParent().getName()).isEqualTo("2022-12-25");
-    Instant thirdRecordInstant = parseDateTime("2022-12-26T12:20:22.3344Z");
-    assertThat(writer3.getPath().getName()).isEqualTo(createExpectedFileName(thirdRecordInstant));
-    assertThat(writer3.getPath().getParent().getName()).isEqualTo("2022-12-26");
-  }
-
-  @Test
-  public void shouldRollover_returnsFalseIfIsNotAfterRolloverDate() throws Exception {
-    TickableFixedClock clock = new TickableFixedClock(parseDateTime("2022-12-25T12:00:00.00Z"));
-    DatePartitionedRecordsWriterFactory datePartitionedRecordsWriterFactory =
-        createRecordsWriterFactory(clock);
-
-    // Act
-    clock.tick(Duration.ofMinutes(25));
-    boolean shouldRollover = datePartitionedRecordsWriterFactory.shouldRollover();
-
-    // Assert
-    assertThat(shouldRollover).isFalse();
-  }
-
-  @Test
-  public void shouldRollover_returnsTrueForNextDate() throws Exception {
-    TickableFixedClock clock = new TickableFixedClock(parseDateTime("2022-12-25T12:00:00.00Z"));
-    DatePartitionedRecordsWriterFactory datePartitionedRecordsWriterFactory =
-        createRecordsWriterFactory(clock);
+    String expectedFileName = "dwhassessment_2022-12-25T120022.3344_a665f132.avro";
```

Comment:
I can see in line 60 in `DatePartitionedRecordsWriterFactory` that the fraction of the second has min and max of the number of decimal places defined:
```
          .appendFraction(NANO_OF_SECOND, 0, 9, true)
```
https://github.com/google/hadoop-migration-assessment-tools/blob/main/src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedRecordsWriterFactory.java#L60

Is there a reason, why the precision of the fraction of the second has 4 decimal places and not e.g. 3 (as for milliseconds) or 6 (as for microseconds)? Is the number of decimal places normalized somewhere else (besides the formatter in DatePartitionedRecordsWriterFactory), so that the filename length doesn't change because of the differences in the number of decimal places across multiple files?

Reply:
Fraction is not normalized, i.e. it could be any number within defined boundaries. In the test instance logs, I've seen only precion up to 3 decimals. I left precision up to nanoseconds, as I'm not sure which resolution other system give.